libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -0,0 +1,204 @@
+			if (m_currentFunctionMemorySlots->count(param.name))
+			{
+				// TODO: we cannot handle function parameters yet.
+				m_currentFunctionMemorySlots = nullptr;
+				break;
+			}
+	}
+	else
+		m_currentFunctionMemorySlots = nullptr;
+	ASTModifier::operator()(_functionDefinition);
+	m_currentFunctionMemorySlots = saved;
+}
+
+void StackToMemoryMover::operator()(Block& _block)
+{
+	using OptionalStatements = std::optional<vector<Statement>>;
+	if (!m_currentFunctionMemorySlots)
+	{
+		ASTModifier::operator()(_block);
+		return;
+	}
+	auto containsVariableNeedingEscalation = [&](auto const& _variables) {
+		return util::contains_if(_variables, [&](auto const& var) {
+			return m_currentFunctionMemorySlots->count(var.name);
+		});
+	};
+	auto rewriteAssignmentOrVariableDeclaration = [&](
+		langutil::SourceLocation const& _loc,
+		auto const& _variables,
+		std::unique_ptr<Expression> _value
+	) -> std::vector<Statement> {
+		if (_variables.size() == 1)
+		{
+			std::vector<Statement> result;
+			appendMemoryStore(
+				result,
+				_loc,
+				getMemoryOffset(_variables.front().name),
+				_value ? *std::move(_value) : Literal{_loc, LiteralKind::Number, "0"_yulstring, {}}
+			);
+			return result;
+		}
+
+		VariableDeclaration tempDecl{_loc, {}, std::move(_value)};
+		vector<Statement> memoryAssignments;
+		vector<Statement> variableAssignments;
+		for (auto& var: _variables)
+		{
+			YulString tempVarName = m_nameDispenser.newName(var.name);
+			tempDecl.variables.emplace_back(TypedName{var.location, tempVarName, {}});
+
+			if (m_currentFunctionMemorySlots->count(var.name))
+				appendMemoryStore(memoryAssignments, _loc, getMemoryOffset(var.name), Identifier{_loc, tempVarName});
+			else if constexpr (std::is_same_v<std::decay_t<decltype(var)>, Identifier>)
+				variableAssignments.emplace_back(Assignment{
+					_loc, { Identifier{var.location, var.name} },
+					make_unique<Expression>(Identifier{_loc, tempVarName})
+				});
+			else
+				variableAssignments.emplace_back(VariableDeclaration{
+					_loc, {std::move(var)},
+					make_unique<Expression>(Identifier{_loc, tempVarName})
+				});
+		}
+		std::vector<Statement> result;
+		result.emplace_back(std::move(tempDecl));
+		std::reverse(memoryAssignments.begin(), memoryAssignments.end());
+		result += std::move(memoryAssignments);
+		std::reverse(variableAssignments.begin(), variableAssignments.end());
+		result += std::move(variableAssignments);
+		return result;
+	};
+
+	util::iterateReplacing(
+		_block.statements,
+		[&](Statement& _statement)
+		{
+			auto defaultVisit = [&]() { ASTModifier::visit(_statement); return OptionalStatements{}; };
+			return std::visit(util::GenericVisitor{
+				[&](Assignment& _assignment) -> OptionalStatements
+				{
+					if (!containsVariableNeedingEscalation(_assignment.variableNames))
+						return defaultVisit();
+					visit(*_assignment.value);
+					auto loc = _assignment.location;
```

Comment:
Why not inline `loc` in the next line? `_assignment` itself isn't moved, only `_assignment.value`

Reply:
True - it may have been moved earlier or I may just have wanted to keep the line shorter this way - but I'll move and introduce line breaks.